### DISCUSSION TITLE
Modifying the querying strategy on a per-query basis

### DISCUSSION THREAD

**Discussion by ghiculescu** [CanCanCommunity/cancancan#668] [Ideas]
In https://github.com/CanCanCommunity/cancancan/pull/655 I added `CanCan.accessible_by_strategy`, which allows you to toggle between different ways `accessible_by` builds queries that contain joins.

@coorasse and @madmax were keen for a way to customise this on a per-query basis. I was hesitant to do that, because it felt like it would add a lot of complexity to the PR and what I had was good enough for my use case. That said, I think it is worth discussing for others, so I'd like to settle on an API first.

## Option 1: new arg

```ruby
CanCan.accessible_by_strategy = :subquery
User.accessible_by(current_ability) # uses subquery strategy
User.accessible_by(current_ability, strategy: :left_joins) # uses left_joins
User.accessible_by(current_ability, strategy: :left_joins) # uses left_joins
User.accessible_by(current_ability, strategy: :left_joins) # uses left_joins
User.accessible_by(current_ability) # uses subquery
```

## Option 2: yielding a scope

```ruby
CanCan.accessible_by_strategy = :subquery
User.accessible_by(current_ability) # uses :subquery

CanCan.with_accessible_by_strategy(:left_joins) do
  # all of these use left_joins
  User.accessible_by(current_ability)
  User.accessible_by(current_ability)
  User.accessible_by(current_ability)
end

User.accessible_by(current_ability) # uses :subquery
```

-------------------------

Thinking this through, I suspect the easiest way to implement option 1 would be to build option 2 and then use it internally. I'm happy to do that, but before I do, would like feedback on either / both of these APIs, as well as thoughts on if there's any better approaches out there.

**Comment by coorasse** [comment]
I don't have much to add since I agree with you 100%. I think I'd normally use option 1, because is less code to write but I'd love to have also option 2 available and, as you said, option 2 could anyway be the internal implementation of option 1. Really nothing to add from my side.

**Comment by ghiculescu** [comment]
https://github.com/CanCanCommunity/cancancan/pull/675
